### PR TITLE
Reader implementation

### PR DESCRIPTION
**Pull request by carlotacb**

### CODE REVIEW

File: package.go [hunk 1]
```diff
@@ -180,7 +180,77 @@ func (c *contentTypes) addDefault(extension, contentType string) {
...
 }
 
-type corePropertiesXML struct {
+func (c *contentTypes) findType(name string) (string, error) {
+	if t, ok := c.overrides[strings.ToUpper(name)]; ok {
+		return t, nil
+	}
+	ext := filepath.Ext(name)
+	if ext != "" {
+		if t, ok := c.defaults[ext[1:]]; ok {
```
Comment (qmuntal) [inline package.go]: what if extension is upper case and default is lower case? The spec says the comparations should be done case insensitive. Is there a unit test for it?

File: reader.go [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+func newReader(a archive) (*Reader, error) {
+	r := &Reader{p: newPackage(), r: a}
+	if err := r.loadPackage(); err != nil {
+		return nil, err
+	}
+	return r, nil
+}
+
+func (r *Reader) loadPackage() error {
+	files := r.r.Files()
```
Comment (qmuntal) [inline reader.go]: This line can be moved after the next error checking

File: reader.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+		if strings.EqualFold(fileName, contentTypesName) || isRelationshipURI(fileName) || strings.HasSuffix(fileName, "/") {
+			continue
+		}
+		if strings.EqualFold(fileName, r.Properties.PartName) {
+			cp, err := r.loadCoreProperties(file)
+			if err != nil {
+				return err
+			}
+			r.Properties = *cp
+			continue
```
Comment (qmuntal) [inline reader.go]: remove continue and add and else case

File: reader.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+		r.p.add(part)
+	}
+	r.p.contentTypes = *ct
+	return nil
+}
+
+func (r *Reader) loadPartPropierties() (*contentTypes, *relationshipsPart, error) {
+	var ct *contentTypes
+	rels := new(relationshipsPart)
+	var err error
```
Comment (qmuntal) [inline reader.go]: move declaration inside the for loop

File: reader.go [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+	files := r.r.Files()
```
Comment (qmuntal) [inline reader.go]: unecessary variable, can be rewritted as range r.r.Files()

File: reader.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+	for _, file := range files {
+		name := "/" + file.Name()
+		if strings.EqualFold(name, contentTypesName) {
+			ct, err = r.loadContentType(file)
+			if err != nil {
```
Comment (qmuntal) [inline reader.go]: move err check after the if/elseif to unify all the following error checks

File: reader.go [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+				return nil, nil, err
+			}
+		} else if isRelationshipURI(name) {
+			if strings.EqualFold(name, packageRelName) {
+				err = r.loadPackageRelationships(file)
+				if err != nil {
```
Comment (qmuntal) [inline reader.go]: move err check after the if/elseif to unify all the following error checks

File: reader.go [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+					return nil, nil, err
+				}
+			} else {
+				err := loadRelationships(file, rels)
+				if err != nil {
```
Comment (qmuntal) [inline reader.go]: move err check after the if/elseif to unify all the following error checks

File: reader.go [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+		return err
+	}
+	rls, err := decodeRelationships(reader)
+	if err != nil {
+		return err
+	}
+	r.Relationships = rls
+	for _, rel := range rls {
+		if strings.EqualFold(rel.Type, corePropsRel) {
+			r.Properties.PartName = rel.TargetURI
```
Comment (qmuntal) [inline reader.go]: should the loop stop if this happen?

File: package.go (hunk 1 above)
Comment (carlotacb) [reply package.go]: I've corrected and added a test for this case, that also check the override case insensitive.

File: reader.go (hunk 9 above)
Comment (carlotacb) [reply reader.go]: break added

File: reader.go (hunk 7 above)
Comment (carlotacb) [reply reader.go]: moved to the end of the for

File: reader.go (hunk 6 above)
Comment (carlotacb) [reply reader.go]: moved to the end of the for

File: reader.go (hunk 5 above)
Comment (carlotacb) [reply reader.go]: deleted variable

File: reader.go (hunk 4 above)
Comment (carlotacb) [reply reader.go]: moved err declaration inside the loop

File: reader.go (hunk 3 above)
Comment (carlotacb) [reply reader.go]: corrected

File: reader.go (hunk 2 above)
Comment (carlotacb) [reply reader.go]: moved

**Pull request merged**